staging/src/k8s.io/apiserver/pkg/server/filters/cors.go

```diff
@@ -38,44 +38,94 @@ func WithCORS(handler http.Handler, allowedOriginPatterns []string, allowedMetho
-					break
-				}
-			}
-			if allowed {
-				w.Header().Set("Access-Control-Allow-Origin", origin)
-				// Set defaults for methods and headers if nothing was passed
-				if allowedMethods == nil {
-					allowedMethods = []string{"POST", "GET", "OPTIONS", "PUT", "DELETE", "PATCH"}
-				}
-				if allowedHeaders == nil {
-					allowedHeaders = []string{"Content-Type", "Content-Length", "Accept-Encoding", "X-CSRF-Token", "Authorization", "X-Requested-With", "If-Modified-Since"}
-				}
-				if exposedHeaders == nil {
-					exposedHeaders = []string{"Date"}
-				}
-				w.Header().Set("Access-Control-Allow-Methods", strings.Join(allowedMethods, ", "))
-				w.Header().Set("Access-Control-Allow-Headers", strings.Join(allowedHeaders, ", "))
-				w.Header().Set("Access-Control-Expose-Headers", strings.Join(exposedHeaders, ", "))
-				w.Header().Set("Access-Control-Allow-Credentials", allowCredentials)
-
-				// Stop here if its a preflight OPTIONS request
-				if req.Method == "OPTIONS" {
-					w.WriteHeader(http.StatusNoContent)
-					return
-				}
-			}
+		if len(origin) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+		if !isOriginAllowed(origin, allowedOriginPatternsREs) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		w.Header().Set("Access-Control-Allow-Origin", origin)
+		w.Header().Set("Access-Control-Allow-Methods", allowMethodsResponseHeader)
+		w.Header().Set("Access-Control-Allow-Headers", allowHeadersResponseHeader)
+		w.Header().Set("Access-Control-Expose-Headers", exposeHeadersResponseHeader)
+		w.Header().Set("Access-Control-Allow-Credentials", allowCredentials)
+
+		// Stop here if its a preflight OPTIONS request
+		if req.Method == "OPTIONS" {
+			w.WriteHeader(http.StatusNoContent)
+			return
 		}
+
 		// Dispatch to the next handler
 		handler.ServeHTTP(w, req)
 	})
 }
 
+// isOriginAllowed returns true if the given origin header in the
+// request is allowed CORS.
+//
+// From https://www.rfc-editor.org/rfc/rfc6454#page-13
+//
+//	 a) The origin header can contain host and/or port
+//			serialized-origin   = scheme "://" host [ ":" port ]
+//
+//	 b) In some cases, a number of origins contribute to causing the user
+//	 agents to issue an HTTP request.  In those cases, the user agent MAY
+//	 list all the origins in the Origin header field. For example, if the
+//	 HTTP request was initially issued by one origin but then later
+//	 redirected by another origin, the user agent MAY inform the server
+//	 that two origins were involved in causing the user agent to issue the
+//	 request
+//			origin-list = serialized-origin *( SP serialized-origin )
+func isOriginAllowed(originHeader string, allowedOriginPatternsREs []*regexp.Regexp) bool {
+	// the origin header may contain multiple origins
+	origins := strings.Split(originHeader, " ")
+	for _, origin := range origins {
+		if len(origin) == 0 {
+			continue
+		}
+		for _, re := range allowedOriginPatternsREs {
+			if re.MatchString(origin) {
+				return true
```

Comment:
is it accurate that matching one origin in a multi-origin request means we should allow the whole chain? I'm not familiar enough with how multi-origin requests work to know if this is right or wrong, but it seems like this would open the door to an allowed origin of `https://good.example.com` permitting a multi-origin chain of `https://good.example.com https://bad.example.com`

Reply:
@liggitt  After diffing a bit, this is what the RFC on says about multiple origins in the  `Origin` header

> In some cases, a number of origins contribute to causing the user
   agents to issue an HTTP request.  In those cases, the user agent MAY
   list all the origins in the Origin header field.  For example, if the
   HTTP request was initially issued by one origin but then later 
> redirected by another origin, the user agent MAY inform the server
   that two origins were involved in causing the user agent to issue the
   request.

So theoretically it is possible to have multiple origins in the `Origin` header, then I looked at the CORS spec here:
https://www.w3.org/TR/2020/SPSD-cors-20200602/#access-control-allow-origin-response-header

> Access-Control-Allow-Origin = "Access-Control-Allow-Origin" ":" [origin-list-or-null](http://tools.ietf.org/html/rfc6454#section-7.1) | "*"

> In practice the [origin-list-or-null](http://tools.ietf.org/html/rfc6454#section-7.1) production is more constrained. Rather than allowing a space-separated list of [origins](http://tools.ietf.org/html/rfc6454#section-4), it is either a single [origin](http://tools.ietf.org/html/rfc6454#section-4) or the string "null".

> If the response includes zero or more than one [Access-Control-Allow-Origin](https://www.w3.org/TR/2020/SPSD-cors-20200602/#http-access-control-allow-origin) header values, return fail and terminate this algorithm

I can think of the following options if the `Origin` header contains multiple origins and CORS is enabled:
- A: do nothing: 
   - if the regexp matches (in the unlikely case) the CORS filter will include the origins in the `Access-Control-Allow-Origin` response header, and the browser will probably error out. 
   - do nothing: if the regexp does not match - did the user expect a certain origin from the `Origin` header to match the regular expression (today we use `req.Header.Get("Origin")` and that returns the first origin. 
- B: if the `Origin` header has multiple origins we don't send CORS response headers (we can send a Warning to the client)
- C: if the `Origin` header has multiple origins, we run the regular expression(s) against each origin and send the first matching origin via `Access-Control-Allow-Origin`

depending on how we want to approach it, I will open a separate PR, this PR seems orthogonal to this issue.

Reply:
I added a test in https://github.com/kubernetes/kubernetes/pull/115242 to understand the current behavior with multiple origins in `Origin` header(s)